frontends/parsers/p4/p4parser.ypp

```diff
@@ -459,10 +459,14 @@ annotation
           $$ = new IR::Annotation(@1, *$2, body); }
     | "@" name "(" annotationBody ")"
         { $$ = new IR::Annotation(@1, *$2, *$4); }
+    | "@" name "[" expressionList "]"
+        { $$ = new IR::Annotation(@1, *$2, *$4, true); }
+    | "@" name "[" kvList "]"
+        { $$ = new IR::Annotation(@1, *$2, *$4, true); }
```

Comment:
I though we would go for a enum with 3 possible values: `UNSTRUCTURED`, `EXPRESSION_LIST`, `KV_LIST`, but I see that you opted for a boolean. I think it makes sense because for every structured annotation you set `needsParsing` to `true`, which means that exactly one of these fields will not be empty:
```
    /// Annotations that are simple expressions
    inline Vector<Expression> expr;

    /// Annotations described as key-value pairs
    inline IndexedVector<NamedExpression> kv;
```
unless the annotation body is empty (`@foo[]`), in which case both of these vectors will be empty.

Please let me know if I got this wrong. The P4Info-generation code needs to be able to "easily" distinguish between both cases.

Reply:
If it is structured then only one of the two can be non-empty.
If both are empty it's the kv that is assumed to be used.
needsParsing should be false for structured annotations. It can only even be true for unstructured annotations. I don't think you need to look at it.
I can add a getter that would return one of these three if that makes your job easier.